src/Executor/Runner/Adapter.php

```diff
@@ -0,0 +1,194 @@
+     * @param string $image
+     * @param string $source
+     * @param string $entrypoint
+     * @param mixed $variables
+     * @param float $cpus
+     * @param int $memory
+     * @param string $version
+     * @param string $runtimeEntrypoint
+     * @param bool $logging
+     * @param string $restartPolicy
+     * @param Table $activeRuntimes
+     * @param Log $log
+     * @return mixed
+     */
+    abstract public function createExecution(
+        string $runtimeId,
+        ?string $payload,
+        string $path,
+        string $method,
+        mixed $headers,
+        int $timeout,
+        string $image,
+        string $source,
+        string $entrypoint,
+        mixed $variables,
+        float $cpus,
+        int $memory,
+        string $version,
+        string $runtimeEntrypoint,
+        bool $logging,
+        string $restartPolicy,
+        Table $activeRuntimes,
+        Log $log
+    ): mixed;
+
+    protected function getStorageDevice(string $root): Device
+    {
+        $connection = Http::getEnv('OPR_EXECUTOR_CONNECTION_STORAGE', '');
+
+        if (!empty($connection)) {
+            $acl = 'private';
+            $device = Storage::DEVICE_LOCAL;
+            $accessKey = '';
+            $accessSecret = '';
+            $bucket = '';
+            $region = '';
+
+            try {
+                $dsn = new DSN($connection);
+                $device = $dsn->getScheme();
+                $accessKey = $dsn->getUser() ?? '';
+                $accessSecret = $dsn->getPassword() ?? '';
+                $bucket = $dsn->getPath() ?? '';
+                $region = $dsn->getParam('region');
+            } catch (\Exception $e) {
+                Console::warning($e->getMessage() . 'Invalid DSN. Defaulting to Local device.');
+            }
+
+            switch ($device) {
+                case Storage::DEVICE_S3:
+                    return new S3($root, $accessKey, $accessSecret, $bucket, $region, $acl);
+                case STORAGE::DEVICE_DO_SPACES:
+                    return new DOSpaces($root, $accessKey, $accessSecret, $bucket, $region, $acl);
+                case Storage::DEVICE_BACKBLAZE:
+                    return new Backblaze($root, $accessKey, $accessSecret, $bucket, $region, $acl);
+                case Storage::DEVICE_LINODE:
+                    return new Linode($root, $accessKey, $accessSecret, $bucket, $region, $acl);
+                case Storage::DEVICE_WASABI:
+                    return new Wasabi($root, $accessKey, $accessSecret, $bucket, $region, $acl);
+                case Storage::DEVICE_LOCAL:
+                default:
+                    return new Local($root);
+            }
+        } else {
+            switch (strtolower(Http::getEnv('OPR_EXECUTOR_STORAGE_DEVICE', Storage::DEVICE_LOCAL) ?? '')) {
+                case Storage::DEVICE_LOCAL:
+                default:
+                    return new Local($root);
+                case Storage::DEVICE_S3:
```

Comment:
would be a good chance to migrate this to DSN format like we do for databases. Will also make it easier to handle any new adapters and maintenance in general.